### PR TITLE
kittest

### PR DESCRIPTION
**Pull request by lucasmerlin**
Initial kittest implementation

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -24,13 +24,15 @@ default = []
...
+accesskit = "0.16.0"
+parking_lot = "0.12"
 
 [dev-dependencies]
-
+egui = { version = "0.28.1", features = ["accesskit"] }
 
 [patch.crates-io]
-
+egui = { git = "https://github.com/emilk/egui"}
```
Comment (emilk) [inline Cargo.toml]: You should specify a specific `rev` here in order to get predicable builds (important when for check we check out old revisions of `kittest`, e.g. for `git bisect`)

File: examples/basic_integration.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
+use egui::{CentralPanel, Context, RawInput};
```
Comment (emilk) [inline examples/basic_integration.rs]: Add a `//! Docstring` explaining what this example demonstrates

File: src/event.rs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+use accesskit::Vec2;
+
+pub enum Event {
+    ActionRequest(accesskit::ActionRequest),
+    Simulated(SimulatedEvent),
+}
+
+pub enum SimulatedEvent {
```
Comment (emilk) [inline src/event.rs]: docstrings!

```suggestion
pub enum Event {
    ActionRequest(accesskit::ActionRequest),
    Simulated(SimulatedEvent),
}

pub enum SimulatedEvent {
```

File: src/node.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
+use crate::event::{Event, SimulatedEvent};
+use crate::query::Queryable;
+use crate::tree::EventQueue;
+use accesskit::{ActionRequest, Vec2};
+use std::fmt::{Debug, Formatter};
+use std::ops::Deref;
+
+pub struct Node<'tree> {
```
Comment (emilk) [inline src/node.rs]: docstring

File: src/query.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        panic!(
+            "Found two or more nodes matching the query: {:?} {:?}",
+            first.node().id(),
+            second.node().id(),
+        );
+    }
+    result
+}
+
+pub trait Queryable<'tree, 'node> {
```
Comment (emilk) [inline src/query.rs]: docstring on the type and all the members!

File: src/tree.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+use crate::event::Event;
+use crate::query::Queryable;
+use crate::Node;
+use accesskit::TreeUpdate;
+use parking_lot::Mutex;
+
+pub struct Tree {
```
Comment (emilk) [inline src/tree.rs]: docstring!

File: src/tree.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+impl Tree {
+    pub fn new(update: TreeUpdate) -> Self {
+        Self {
+            tree: accesskit_consumer::Tree::new(update, true),
+            queued_events: Mutex::new(Vec::new()),
+        }
+    }
+
+    pub fn update(&mut self, update: accesskit::TreeUpdate) {
```
Comment (emilk) [inline src/tree.rs:25]: docstring!

File: src/tree.rs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        self.tree.update(update);
+    }
+
+    pub fn root(&self) -> Node<'_> {
+        self.node()
+    }
+
+    pub fn take_events(&self) -> Vec<Event> {
+        self.queued_events.lock().drain(..).collect()
```
Comment (emilk) [inline src/tree.rs:36]: Why is this `&self` and not `&mut self` (i.e. why is there a Mutex)?

**Review by emilk — APPROVED** [review]

File: src/tree.rs (hunk 8 above)
Comment (lucasmerlin) [reply src/tree.rs:36]: I wanted the query functions to take &self, so I don't think a Node can hold a mutable reference to the event queue.
If the nodes would hold mutable references I think you would get problems when querying multiple nodes. 

I think I could replace this with a RefCell though.

File: src/event.rs (hunk 3 above)
Comment (lucasmerlin) [reply src/event.rs]: What suggestion did you want to make? There are no changes

File: Cargo.toml (hunk 1 above)
Comment (lucasmerlin) [reply Cargo.toml]: Doesn't the Cargo.lock take care of that? 🤔

Comment (emilk) [reply Cargo.toml]: Yeah, good point 👍

File: src/event.rs (hunk 3 above)
Comment (emilk) [reply src/event.rs]: That was a mistake - but docstrings :)

File: src/event.rs [hunk 9]
```diff
@@ -0,0 +1,241 @@
+use accesskit::Vec2;
+
+/// Kittest Event.
+pub enum Event {
+    /// An ActionRequest event. When using an application these would be generated by the
+    /// screen reader.
+    ActionRequest(accesskit::ActionRequest),
+    /// A Simulated event, e.g. clicks or typing.
```
Comment (emilk) [inline src/event.rs]: nit: there is some weird casing going on here

```suggestion
/// Kittest event sent to the GUI application.
pub enum Event {
    /// An [`accesskit::ActionRequest`] event. When using an application these would be generated by the
    /// screen reader.
    ActionRequest(accesskit::ActionRequest),
    
    /// A simulated event, e.g. clicks or typing.
```

File: src/event.rs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    F33,
+    F34,
+    F35,
+    // When adding keys, remember to also update:
+    // * crates/egui-winit/src/lib.rs
+    // * Key::ALL
+    // * Key::from_name
+    // You should test that it works using the "Input Event History" window in the egui demo app.
+    // Make sure to test both natively and on web!
+    // Also: don't add keys last; add them to the group they best belong to.
```
Comment (emilk) [inline src/event.rs]:
```suggestion
```
```suggestion
    // Don't always add keys last; add them to the group they best belong to.
```

File: src/filter.rs [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    /// Default is false.
+    pub fn include_labels(mut self) -> Self {
+        self.include_labels = true;
+        self
+    }
+
+    /// Filter by a custom predicate.
+    pub fn predicate(mut self, predicate: impl Fn(&Node<'_>) -> bool + 'a) -> Self {
+        self.predicate = Some(Box::new(predicate));
+        self.had_predicate = true;
```
Comment (emilk) [inline src/filter.rs:102]: Why is `had_predicate` needed? Isn't it the same as `self.predicate.is_some()`?

File: src/node.rs [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+                state,
+            }));
+        });
+    }
+
+    /// Focus the node and type the given text
+    pub fn type_text(&self, text: &str) {
+        self.focus();
+        self.event(Event::Simulated(SimulatedEvent::Ime(text.to_owned())));
+    }
```
Comment (emilk) [inline src/node.rs]: nit: slightly more ergonomic

```suggestion
    pub fn type_text(&self, text: impl Into<String>) {
        self.focus();
        self.event(Event::Simulated(SimulatedEvent::Ime(text.into())));
    }
```

**Review by emilk — COMMENTED** [review]
Looks great! Make sure to fix those clippy warnings though

**Review by emilk — APPROVED** [review]
Looks great! Make sure to fix those clippy warnings though

File: src/filter.rs (hunk 11 above)
Comment (lucasmerlin) [reply src/filter.rs:102]: It's used in By::debug_clone. This way I don't have to clone the predicate function (which wouldn't be useful in debug prints anyways) but can tell the user that the filter was filtered by a custom predicate.

**Pull request merged**